File: android/app/src/main/kotlin/net/eduverseapp/platform/upload/UploadStore.kt
package net.eduverseapp.platform.upload

import android.content.Context
import org.json.JSONObject
import java.io.File

/**
 * File-backed persistence for the native upload pipeline. Two areas:
 *  - `pending/<jobId>.json`  : the [UploadJobData] for a job not yet finished.
 *    Written when enqueued, deleted when the job reaches a terminal state.
 *  - `results/<jobId>.json`  : the terminal [UploadResult] for the Dart side to
 *    read and reconcile. Deleted once Dart acknowledges via clearResult.
 *
 * Because both live in the app's private filesDir, they survive app kill and
 * device reboot, which is what lets WorkManager resume the queue unattended.
 */
class UploadStore(context: Context) {

    private val pendingDir = File(context.filesDir, "eduverse_upload/pending").apply { mkdirs() }
    private val resultsDir = File(context.filesDir, "eduverse_upload/results").apply { mkdirs() }

    fun savePending(job: UploadJobData) {
        File(pendingDir, "${job.jobId}.json").writeText(job.toJson().toString())
    }

    fun loadPending(jobId: String): UploadJobData? {
        val f = File(pendingDir, "$jobId.json")
        if (!f.exists()) return null
        return try {
            UploadJobData.fromJson(JSONObject(f.readText()))
        } catch (e: Exception) {
            null
        }
    }

    fun deletePending(jobId: String) {
        File(pendingDir, "$jobId.json").delete()
    }

    fun saveResult(result: UploadResult) {
        File(resultsDir, "${result.jobId}.json").writeText(result.toJson().toString())
    }

    fun allResults(): List<JSONObject> {
        return resultsDir.listFiles()
            ?.filter { it.extension == "json" }
            ?.mapNotNull { runCatching { JSONObject(it.readText()) }.getOrNull() }
            ?: emptyList()
    }

    fun clearResult(jobId: String) {
        File(resultsDir, "$jobId.json").delete()
    }

    // MARK: - Progress tracking (read by Dart poll cycle for real-time UI)

    private val progressDir = File(context.filesDir, "eduverse_upload/progress").apply { mkdirs() }

    fun saveProgress(jobId: String, pct: Int) {
        try {
            File(progressDir, "$jobId.json").writeText("{\"progress\":$pct}")
        } catch (_: Exception) {}
    }

    fun loadProgress(jobId: String): Int? {
        val f = File(progressDir, "$jobId.json")
        if (!f.exists()) return null
        return try {
            JSONObject(f.readText()).optInt("progress", -1).let { if (it < 0) null else it }
        } catch (_: Exception) { null }
    }

    fun deleteProgress(jobId: String) {
        try { File(progressDir, "$jobId.json").delete() } catch (_: Exception) {}
    }
}

File: ios/Runner/UploadStore.swift
import Foundation

/// File-based persistence for the iOS native upload pipeline.
/// Mirrors the Android `UploadStore` pattern.
///
/// Directories:
///   Library/Caches/eduverse_upload/pending/<jobId>.json — UploadJobData
///   Library/Caches/eduverse_upload/results/<jobId>.json  — UploadResult
class UploadStore {
    private let pendingDir: URL
    private let resultsDir: URL
    private let fileManager = FileManager.default

    init() {
        let base = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first!
            .appendingPathComponent("eduverse_upload", isDirectory: true)
        pendingDir = base.appendingPathComponent("pending", isDirectory: true)
        resultsDir = base.appendingPathComponent("results", isDirectory: true)
        try? fileManager.createDirectory(at: pendingDir, withIntermediateDirectories: true)
        try? fileManager.createDirectory(at: resultsDir, withIntermediateDirectories: true)
    }

    // MARK: - Pending jobs

    func savePending(_ job: UploadJobData) {
        let url = pendingDir.appendingPathComponent("\(job.jobId).json")
        if let data = try? JSONSerialization.data(withJSONObject: job.toDict()) {
            try? data.write(to: url)
        }
    }

    func loadPending(_ jobId: String) -> UploadJobData? {
        let url = pendingDir.appendingPathComponent("\(jobId).json")
        guard let data = try? Data(contentsOf: url),
              let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return UploadJobData(from: dict)
    }

    func deletePending(_ jobId: String) {
        let url = pendingDir.appendingPathComponent("\(jobId).json")
        try? fileManager.removeItem(at: url)
    }

    // MARK: - Results

    func saveResult(_ result: UploadResult) {
        let url = resultsDir.appendingPathComponent("\(result.jobId).json")
        if let data = try? JSONSerialization.data(withJSONObject: result.toDict()) {
            try? data.write(to: url)
        }
    }

    func allResults() -> [[String: Any]] {
        guard let files = try? fileManager.contentsOfDirectory(
            at: resultsDir, includingPropertiesForKeys: nil
        ) else { return [] }
        return files.compactMap { url in
            guard url.pathExtension == "json",
                  let data = try? Data(contentsOf: url),
                  let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            else { return nil }
            return dict
        }
    }

    func clearResult(_ jobId: String) {
        let url = resultsDir.appendingPathComponent("\(jobId).json")
        try? fileManager.removeItem(at: url)
    }

    // MARK: - Counts

    var pendingCount: Int {
        (try? fileManager.contentsOfDirectory(at: pendingDir, includingPropertiesForKeys: nil)
            .filter { $0.pathExtension == "json" }.count) ?? 0
    }
}
